Implement reflection support for Symbolic Extended Existential types.

adrian-prantl opened the pull request:
This patch add a new SymbolicExtendedExistentialTypeRef kind wires it up in TypeDecoder, TypeRefBuilder, TypeLowering, and ASTDemangler. This is tested indirectly via the matching LLDB commit.

EDIT: This is now ready for review!

adrian-prantl commented on stdlib/public/RemoteInspection/TypeLowering.cpp:
```diff
@@ -1258,6 +1259,8 @@ class ExistentialTypeInfoBuilder {
     }
   }
 
+  void examineShape() {}
```
Reviewing my own patch here: This is dead code.

adrian-prantl commented on include/swift/Remote/MetadataReader.h:
```diff
@@ -886,6 +886,77 @@ class MetadataReader {
+    StoredPointer addr = shapeAddress.getResolvedAddress().getAddressData();
+    // Non-unique shapes start with an offset to a cache variable.
+    if (non_unique)
+      addr += 4;
```
@rjmccall Non-unique shapes have an extra field at the beginning:
https://github.com/swiftlang/swift/blob/a654df7caaff587aaf548fa5ef42b82ce35966bb/lib/IRGen/GenMeta.cpp#L7748

I'm adding the offset here, so the existing code remains unchanged, but maybe it would be better to teach readShape about this? It uses the address as a cache entry.

adrian-prantl replied:
To summarize an offline conversation: Shape symbols have and extra field, but all metadata pointers point directly to the shape data after that field, so this is the desired behavior.

adrian-prantl commented on include/swift/Remote/MetadataReader.h:
```diff
@@ -1081,79 +1152,27 @@ class MetadataReader {
-      }
 
+      // FIXME!
+      bool non_unique = false;
```
@rjmccall Similarly, Im wondering if this existing code is currently silently behaving incorrectly for non-unique shapes?

adrian-prantl commented on include/swift/Remote/MetadataReader.h:
```diff
@@ -2148,6 +2166,7 @@ class MetadataReader {
           return nullptr;
         auto shapePtr = stripSignedPointer(signedShapePtr);
 
+        // FIXME: Apply non-unique +4 offset?
```
And this one.

adrian-prantl commented:
test with https://github.com/swiftlang/llvm-project/pull/10915
@swift-ci test

adrian-prantl commented:
test with https://github.com/swiftlang/llvm-project/pull/10915
@swift-ci test

rjmccall commented on include/swift/Remote/MetadataReader.h:
```diff
@@ -1382,38 +1395,42 @@ class MetadataReader {
+                                                         /*nullable*/ false>;
+      constexpr size_t buffer_size = sizeof(ExtendedExistentialTypeShapeFlags) +
+                                     sizeof(RelativeOffset) +
+                                     sizeof(GenericContextDescriptorHeader);
```
Please try:

```
union {
  TargetExtendedExistentialTypeShape<Runtime> shapeHeader;
};
if (!Reader->readBytes(RemoteAddress(address), reinterpret_cast<uint8_t*>(&shapeHeader), sizeof(shapeHeader))
  return nullptr;
```

and then you should be able to just use `shapeHeader.Flags` and `shapeHeader.ReqSigHeader` without any other fuss.

adrian-prantl replied:
That didn't work directly because TargetExtendedExistentialTypeShape has no default constructor, but I updated the patch to do achieve something pretty close.

adrian-prantl commented:
test with https://github.com/swiftlang/llvm-project/pull/10915
@swift-ci test

adrian-prantl committed:
Implement reflection support for Symbolic Extended Existential types.

This patch adds a new SymbolicExtendedExistentialTypeRef kind, and
wires it up in TypeDecoder, TypeRefBuilder, TypeLowering, and
ASTDemangler.

This is tested indirectly via the matching LLDB commit.

adrian-prantl commented:
test with https://github.com/swiftlang/llvm-project/pull/10915
@swift-ci test

adrian-prantl commented:
Rebased on top of @augusto2112 's RemoteAddress patch.

augusto2112 reviewed (approved):
I left a bunch of nits, the one potentially concerning thing I see is one usage of `DefaultAddressSpace` (which may or may not be legitimate).

adrian-prantl commented:
test with https://github.com/swiftlang/llvm-project/pull/10915
@swift-ci test
